### PR TITLE
[FLINK-26126][kafka] fix numRecordsOut metric error

### PR DESCRIPTION
**Pull request by JingGe**
## What is the purpose of the change

We found that the new sink v2 interface will have a wrong numRecordsOut metric for the sink writers. We send a fixed number of records to the source, but the numRecordsOut of the sink continues to increase by the time.

The problem is that both the `SinkWriterOperator` and the `KafkaWriter` are using the same counter metric for counting the outgoing records. Same records sent by the `SinkWriterOperator` to the post topology and written by the `KafkaWriter` to the downstream system will be count twice in the same counter metric.

## Brief change log

  -  define new counters in `SinkWriterMetricGroup`
  -  replace the shared counter used in `KafkaWriter` with new counters
  -  update `KafkaWriterITCase`
  -  enable `KafkaSinkITCase` and `KafkaSinkE2ECase` again

## Verifying this change

This change is already covered by existing tests, such as `KafkaWriterITCase`, `KafkaSinkITCase`, and `KafkaSinkE2ECase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (**yes** / no / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (will be done in a follow-up PR)

## Future Work

- follow-up task for naming consistency: FLINK-26282

### CODE REVIEW

**Comment by JingGe** [comment]
@flinkbot run azure

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSinkWriterMetricGroup.java
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSinkWriterMetricGroup.java:83]: What's up with the the lack of imports?

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSinkWriterMetricGroup.java
Comment (zentol) [inline flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSinkWriterMetricGroup.java:48]: This doesn't really work. You are registering a second counter with the same name of one that already exists, returning you a dummy counter which is never exported. You could just as well create a plain SimpleCounter if that is the intention.

**Review by zentol — CHANGES_REQUESTED** [review]

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSinkWriterMetricGroup.java
Comment (JingGe) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSinkWriterMetricGroup.java:83]: ops, I have changed the IDEA settings for other issues. I will change it back. Thanks.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSinkWriterMetricGroup.java
Comment (zentol) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSinkWriterMetricGroup.java:48]: > Second question, What do you mean "never exported"? Is there anything else should be done to make it be exported? Do we have tech guide somewhere?

With that I mean that the metric will never be passed to a reporter. You can't work your way around that, beyond solving the issue of registering 2 metrics with the same name on the same group.

**Comment by zentol** [comment]
You said the SinkWriterOperator counts outgoing records; where does that happen and why did FLINK-24382 not fix that?

**Comment by zentol** [comment]
aha. The SinkWriterOperator passes data to the sinkWriter (which counts stuff as it writes them to Kafka), but then also sends a commitable to the co-located committer, which also of course counts as a record.

So, question 1: This doesn't seem specific to Kafka but a fundamental flaw in how the SinkWriterOperator works. Why are you only addressing it for Kafka?

Question 2: How did you guys think this should be exposed to the user? What you are running into is an unanswered question w.r.t. metrics, as you have 1 operator writing data to the outside while also writing data to the next operator. Now what is the numRecordsOut for that operator?

**Comment by zentol** [comment]
From what I remember from previous discussions with Arvid on this topic, is that we should have separate metrics for data written from operator to operator (numRecordsOut) and data written to the outside (numRecordsOutExternal?...).

But that's not something I'd be keen on adding after the feature freeze.

File: flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSinkWriterMetricGroup.java
Comment (JingGe) [reply flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/InternalSinkWriterMetricGroup.java]: Thanks for the feedback. Comments have been addressed.

**Comment by JingGe** [comment]
> aha. The SinkWriterOperator passes data to the sinkWriter (which counts stuff as it writes them to Kafka), but then also sends a commitable to the co-located committer, which also of course counts as a record.
> 
> So, question 1: This doesn't seem specific to Kafka but a fundamental flaw in how the SinkWriterOperator works. Why are you only addressing it for Kafka?

Yes, every connector that implements sink v2 should have the same issue. Currently, we only have the issue with Kafka and maybe File. I pushed the solution only for Kafka to fix the urgent bug and get feedback. Once the solution is approved, it will be applied to other connectors.

> 
> Question 2: How did you guys think this should be exposed to the user? What you are running into is an unanswered question w.r.t. metrics, as you have 1 operator writing data to the outside while also writing data to the next operator. Now what is the numRecordsOut for that operator? It's clearly not either/or. If you drop the records sent to the committer then you have an operator (the committer) with 0 input. If you drop the Writer count, well then users complain that we only "wrote 1 record) or something like that.

We will keep both metrics, `numRecordsOut` will be used as it was for the records out to the next operator. A new one called `numRecordsSend` will be used for sending the data to the outside downstream system.

**Comment by JingGe** [comment]
@flinkbot run azure

**Comment by JingGe** [comment]
@flinkbot run azure

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:153]: Please use assertJ for new assertions `assertThat(...).isEqualTo()`

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java:412]: This seems to be some sort of new feature. Can you add it on a separate commit and a test for it?

**Review by fapaul — COMMENTED** [review]
The changes look good I only have some minor comments regarding the test. I think should also update the `SinkMetricsITCase` because it currently still uses the recordsOut metric and not recordsSent.

**Comment by fapaul** [comment]
In the end it probably makes sense to have two/three commits

1. Introduce the new metric with its name and update the `SinkMetricsITCase`
2. Update the kafka metric
3. Add the recordsError metrics

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java
Comment (JingGe) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java:412]: This is on the unhappy path that is normally not covered by ITCase. Do you have any idea on how to test it?

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java
Comment (JingGe) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:153]: This is a JUnit 4 test. It make more sense to do the migration in the separate thread. WDYT?

Comment (fapaul) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:153]: I think the test is already junit 5 but yes probably it makes sense to switch the whole class to use assertJ as separate commit. If you want to do it on this PR fine but also okay if you keep the current assertions.

Comment (JingGe) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:153]: you are right. I will do it in new commit in this PR.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java
Comment (fapaul) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java:412]: One scenario that might work is that you create a second `FlinkKafkaInternalProducer` to fence the writer.

So you would first write some records to kafka with the writer and before calling `prepareCommit` create a FlinkKafkaInternalProducer that uses the same transaction id to writes some records and commits. I suspect that now calling `write` and `prepareCommit` on the writer lead to a `ProducerFencedException`.

Comment (JingGe) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java:412]: excellent!

**Comment by JingGe** [comment]
@flinkbot run azure

**Comment by JingGe** [comment]
@flinkbot run azure

**Comment by JingGe** [comment]
@fapaul Thanks for the effort and feedback. I have addressed all you comments. Please check it again and help merging the PR.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java:266]: Do you really need to expose the recordSerializer and the sink context? I only see them used by the `testNumRecordsOutErrorsCounterMetric` and I think both object are passed to writer during the construction so they are already accessible from the test class.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java
Comment (fapaul) [inline flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java:422]: I think it is not safe to call this here because it is called by the Kafka thread. The counter should only be updated by the mail box thread. In general, I am a bit skeptical about introducing this change because the maximum number of errors is 1 since the pipeline immediately fails after coming to this point.

**Review by fapaul — COMMENTED** [review]
Now that you have introduced metrics that all include the word `Send` I am wondering whether it makes sense to also rename `numRecordsOutErrors` to `numRecordsSendErrors` because out seems to be ambiguous again.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java
Comment (JingGe) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java:422]: The metric is useful for user to be aware of the error even if it only have 1. I will move the call to the mail box thread.

File: flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java
Comment (JingGe) [reply flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java:266]: Thanks for the hint.

**Comment by JingGe** [comment]
> Now that you have introduced metrics that all include the word Send I am wondering whether it makes sense to also rename numRecordsOutErrors to numRecordsSendErrors because out seems to be ambiguous again.

`numRecordsOutErrors` has been used in other places. I am not sure whether it is safe enough to rename it directly. I have nothing again it. I would suggest firing a new follow-up ticket for it. WDYT?

**Review by fapaul — COMMENTED** [review]
> numRecordsOutErrors has been used in other places. I am not sure whether it is safe enough to rename it directly but I have nothing against it. I would suggest firing a new follow-up ticket for it. WDYT? 

I am fine with having it as follow but I would like to fix it before 1.15 because you are introducing the `send` metrics with this PR and it might be confusing. The best way would be probably to deprecate `numRecordsOutError` and introduce `numRecordsSendError`. So we can remove `numRecordsOutError` with 1.16.

I left one small last question regarding the test code.

File: flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java
Comment (JingGe) [reply flink-connectors/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/sink/KafkaWriterITCase.java:238]: Using `contains` will require an additional check for the size. I think it is not necessary to use this call, so removed it.

**Comment by JingGe** [comment]
created follow-up ticket: FLINK-26492

**Comment by JingGe** [comment]
@fapaul would you please help merging this PR? Thanks.

**Pull request merged**
